File: ios/MenCare/Core/Data/Network/NetworkAdminRepository.swift
import Foundation

@MainActor
final class NetworkAdminRepository: AdminRepository {

    private let client: ApiClient

    init(client: ApiClient) {
        self.client = client
    }

    func dashboard(_ period: DashboardPeriod) async throws -> DashboardStats {
        try await apiThrowing {
            // Incassi, occupazione, scontrino medio e clienti inattivi arrivano
            // calcolati: erano la cosa che le due app inventavano piu' liberamente.
            try await client.send(AdminEndpoint.dashboard(period), as: DashboardDto.self).toDomain()
        }
    }

    func notificationSettings() async throws -> NotificationSettings {
        try await apiThrowing {
            // Interruttori e promemoria stanno su due rotte: la schermata li
            // mostra come una cosa sola, quindi si uniscono qui.
            let switches = try await client.send(
                AdminEndpoint.notificationSettings, as: SalonNotificationSettingsDto.self
            )
            let rules = try await client.send(AdminEndpoint.reminderRules, as: ReminderRulesResponseDto.self)
            return NotificationSettings(
                reminders: rules.rules
                    .map { $0.toDomain() }
                    .sorted { $0.hoursBefore > $1.hoursBefore },
                bookingConfirmation: switches.bookingConfirmation,
                cancellationAlert: switches.cancellationAlert,
                lateOperatorAlert: switches.lateOperatorAlert,
                emptyDayPromos: switches.emptyDayPromos
            )
        }
    }

    func updateNotificationSettings(_ settings: NotificationSettings) async -> AppResult<NotificationSettings> {
        await apiResult {
            _ = try await client.send(
                AdminEndpoint.saveNotificationSettings(
                    SalonNotificationSettingsDto(
                        bookingConfirmation: settings.bookingConfirmation,
                        cancellationAlert: settings.cancellationAlert,
                        lateOperatorAlert: settings.lateOperatorAlert,
                        emptyDayPromos: settings.emptyDayPromos
                    )
                ),
                as: SalonNotificationSettingsDto.self
            )

            // I promemoria si creano e si cancellano uno a uno: si confronta la
            // lista in arrivo con quella sul server e si tocca solo la differenza.
            let existing = try await client.send(AdminEndpoint.reminderRules, as: ReminderRulesResponseDto.self).rules
            let wantedHours = settings.reminders.map(\.hoursBefore)
            var remaining = wantedHours

            for rule in existing {
                if let index = remaining.firstIndex(of: rule.hoursBefore) {
                    remaining.remove(at: index)
                } else {
                    _ = try await client.send(AdminEndpoint.removeReminderRule(rule.id))
                }
            }
            for hours in remaining {
                _ = try await client.send(
                    AdminEndpoint.addReminderRule(ReminderRuleWriteDto(hoursBefore: hours)), as: ReminderRuleDto.self
                )
            }
            return try await notificationSettings()
        }
    }

    func campaigns() async throws -> [PushCampaign] {
        try await apiThrowing {
            try await client.send(AdminEndpoint.campaigns, as: CampaignsResponseDto.self)
                .campaigns.map { $0.toDomain() }
        }
    }

    func saveCampaign(_ campaign: PushCampaign) async -> AppResult<PushCampaign> {
        await apiResult {
            // "Invia ora" non e' uno stato che si scrive: si salva la campagna e
            // poi si chiede l'invio, che il server mette in coda (202).
            let sendNow = campaign.status == .sent
            let body = CampaignWriteDto(
                name: campaign.name,
                segment: WireEnum.campaignSegmentWire(campaign.segment),
                title: campaign.title,
                body: campaign.body,
                scheduledAt: campaign.scheduledAt?.isoLocal,
                repeatWeekly: campaign.repeatWeekly,
                sendCap: campaign.sendCap,
                status: sendNow ? "DRAFT" : (campaign.status == .scheduled ? "SCHEDULED" : "DRAFT")
            )
            let saved = try await client.send(
                campaign.id.isEmpty
                    ? AdminEndpoint.createCampaign(body)
                    : AdminEndpoint.updateCampaign(campaign.id, body),
                as: CampaignDto.self
            )
            guard sendNow else { return saved.toDomain() }
            // La risposta dell'invio non interessa alla schermata: conta che sia
            // stata presa in carico.
            _ = try await client.send(AdminEndpoint.sendCampaign(saved.id))
            var sent = saved.toDomain()
            sent.status = .sent
            return sent
        }
    }

    func reachFor(_ segment: CampaignSegment) async throws -> (reachable: Int, size: Int) {
        try await apiThrowing {
            let dto = try await client.send(AdminEndpoint.reach(segment), as: CampaignReachDto.self)
            return (dto.reachable, dto.segmentSize)
        }
    }
}

File: android/core/data/src/main/kotlin/com/devora/mencare/core/data/network/NetworkAdminRepository.kt
package com.devora.mencare.core.data.network

import com.devora.mencare.core.common.AppResult
import com.devora.mencare.core.common.DispatcherProvider
import com.devora.mencare.core.common.onSuccess
import com.devora.mencare.core.data.repository.AdminRepository
import com.devora.mencare.core.model.CampaignSegment
import com.devora.mencare.core.model.CampaignStatus
import com.devora.mencare.core.model.DashboardPeriod
import com.devora.mencare.core.model.DashboardStats
import com.devora.mencare.core.model.NotificationSettings
import com.devora.mencare.core.model.PushCampaign
import com.devora.mencare.core.network.api.AdminApi
import com.devora.mencare.core.network.apiCall
import com.devora.mencare.core.network.dto.CampaignBody
import com.devora.mencare.core.network.dto.CampaignReachDto
import com.devora.mencare.core.network.dto.CreateReminderRuleBody
import com.devora.mencare.core.network.dto.jsonOrNull
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.withContext
import java.time.LocalDateTime
import java.time.temporal.ChronoUnit
import javax.inject.Inject
import javax.inject.Singleton

private val EmptyDashboard = DashboardStats(
    period = DashboardPeriod.WEEK,
    revenueCents = 0,
    revenueTrendPercent = 0,
    appointmentCount = 0,
    noShowPercent = 0.0,
    averageTicketCents = 0,
    operatorOccupancy = emptyList(),
    upcomingDays = emptyList(),
    inactiveClients = emptyList(),
)

/**
 * Area titolare dal backend.
 *
 * I numeri del cruscotto non si ricostruiscono più in app: incassi,
 * occupazione, no-show e scontrino medio sono conti su tutto lo storico, e
 * l'app ne ha in mano solo una manciata di righe. Lo stesso vale per i
 * destinatari raggiungibili di una campagna, che il server conta davvero
 * invece di stimarli.
 */
@Singleton
class NetworkAdminRepository @Inject constructor(
    private val api: AdminApi,
    private val sync: DataSync,
    private val dispatchers: DispatcherProvider,
) : AdminRepository {

    override fun dashboard(period: DashboardPeriod): Flow<DashboardStats> =
        sync.reloading(SyncKeys.DASHBOARD, EmptyDashboard.copy(period = period)) {
            api.dashboard(period.name).toModel()
        }

    /**
     * Gli interruttori e i promemoria sono due rotte distinte ma una sola
     * schermata: si leggono insieme e si scrivono insieme.
     */
    override val notificationSettings: Flow<NotificationSettings> =
        sync.reloading(SyncKeys.SETTINGS, NotificationSettings(emptyList())) {
            coroutineScope {
                val switches = async { api.notificationSettings() }
                val rules = async { api.reminderRules().rules }
                switches.await().toModel(rules.await())
            }
        }

    override suspend fun updateNotificationSettings(settings: NotificationSettings): AppResult<Unit> =
        withContext(dispatchers.io) {
            apiCall {
                api.updateNotificationSettings(settings.toDto())
                syncReminderRules(settings)
            }.onSuccess { sync.invalidate() }
        }

    /**
     * Un promemoria è soltanto "N ore prima": l'insieme delle ore *è* lo stato.
     * Si confrontano quelle, non gli id, perché la schermata crea le righe
     * nuove con un id inventato in locale che il server non ha mai visto.
     */
    private suspend fun syncReminderRules(settings: NotificationSettings) {
        val current = api.reminderRules().rules
        val wanted = settings.reminders.map { it.hoursBefore }.toSet()
        current.filter { it.hoursBefore !in wanted }.forEach { api.removeReminderRule(it.id) }
        val existing = current.map { it.hoursBefore }.toSet()
        (wanted - existing).forEach { api.addReminderRule(CreateReminderRuleBody(it)) }
    }

    override val campaigns: Flow<List<PushCampaign>> =
        sync.reloading(SyncKeys.CAMPAIGNS, emptyList<PushCampaign>()) {
            api.campaigns().campaigns.map { it.toModel() }
        }

    override suspend fun saveCampaign(campaign: PushCampaign): AppResult<PushCampaign> =
        withContext(dispatchers.io) {
            val body = campaign.toBody()
            apiCall {
                val saved = if (campaign.id.isBlank()) {
                    api.createCampaign(body)
                } else {
                    api.updateCampaign(campaign.id, body)
                }
                saved.toModel()
            }.onSuccess { sync.invalidate() }
        }

    override suspend fun reachFor(segment: CampaignSegment): Pair<Int, Int> = withContext(dispatchers.io) {
        val reach = sync.load<CampaignReachDto?>(SyncKeys.CAMPAIGNS, null) { api.campaignReach(segment.name) }
        (reach?.reachable ?: 0) to (reach?.segmentSize ?: 0)
    }
}

private fun PushCampaign.toBody() = CampaignBody(
    name = name,
    segment = segment.name,
    title = title,
    body = body,
    // Orario da muro senza secondi: è la forma che il backend riconosce come
    // "ora scelta dal titolare" e riporta nel fuso del salone.
    scheduledAt = jsonOrNull(scheduledAt?.toMinuteText()),
    repeatWeekly = repeatWeekly,
    sendCap = jsonOrNull(sendCap),
    // Il server accetta solo bozza o programmata: una campagna già inviata non
    // si riporta indietro con un salvataggio.
    status = status.takeIf { it != CampaignStatus.SENT }?.name,
)

private fun LocalDateTime.toMinuteText(): String = truncatedTo(ChronoUnit.MINUTES).toString()
